Mathlib/Control/Monad/AlternativeMonad.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2019 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon
+-/
+import Mathlib.Control.Lawful
+import Mathlib.Control.Monad.Basic
+import Mathlib.Data.Finset.Functor
+
+/-!
+# Laws for Monads with Failure
+
+Definitions for monads that also have an `Alternative` instance while sharing the underlying
+`Applicative` instance, and a class `LawfulAlternative` for types where the failure and monad
+structures are compatible in a natural way. More specifically they satisfy:
+
+* `failure >>= g = failure`
+* `x <|> failure = x`
+* `failure <|> y = y`
+
+`Option`/`OptionT` are the most basic examples, but transformers like `StateT` also preserve
+the lawfullness of this on the underlying monad.
+
+We also include an additional condition for `mapConst`, as it is not necessarilly equal to the
+composition of map with a constant function, and is used in definitions like `success`.
```

Comment:
`LawfulFunctor` ensures that this condition _is_ true, right?

Reply:
Yes I had missed the duplicated naming, removed now